### PR TITLE
Sending-Message-To-Mattermost

### PR DESCRIPTION
**Pull request by badrAZ**

### CODE REVIEW

**Comment by julien-f** [comment]
@pdonias Could you do a first pass review?

File: README.md [hunk 1]
```diff
@@ -12,11 +12,72 @@ Installation of the [npm package](https://npmjs.org/package/xo-server-transport-
...
 
-**TODO**
+Like all other xo-server plugins, it can be configured directly via
+the web iterface, see [the plugin documentation](https://xen-orchestra.com/docs/plugins.html). You can also test the configuration plugin if it works.
+
+### Slack
+
+#### Generate the Webhook
+
+1. Login in your Slack team account.
```
Comment (pdonias) [inline README.md]: "Log in your Slack team account"

File: README.md [hunk 2]
```diff
@@ -12,11 +12,72 @@ Installation of the [npm package](https://npmjs.org/package/xo-server-transport-
...
+
+4. Click on **Add Configuration**.
+
+![Add Configuration](image/DocImg3.png)
+
+5. Choose the default channel and click on **Add Incoming WebHooks integration**.
+
+![Add Incoming WebHooks integration](image/DocImg4.png)
+
+6. Modify the default setting and click on **Save Settings**.
```
Comment (pdonias) [inline README.md]: "default settingS"?

File: src/index.js [hunk 3]
```diff
@@ -1,27 +1,86 @@
...
+      type: 'string',
+      description: 'Channel, private group, or IM channel to send message to.'
+    },
+    username: {
+      type: 'string',
+      description: 'Bot username.'
+    },
+    icon_emoji: {
+      type: 'string',
+      description: 'The bot icone. It can be a slack emoji or an URL image.'
```
Comment (pdonias) [inline src/index.js]: icon*

File: README.md (hunk 2 above)
Comment (badrAZ) [reply README.md]: yes, it's default settings i will fix it.

File: src/index.js [hunk 4]
```diff
@@ -1,27 +1,86 @@
-export const configurationSchema = null
+import Slack from 'slack-node'
+import { promisify } from 'promise-toolbox'
 
-export const testSchema = null
+// ===================================================================
 
-class XoServerTransportNagios {
+const logAndRethrow = error => {
+  console.error('[WARN] plugin transport-slack:', error && error.stack || error)
```
Comment (pdonias) [inline src/index.js:7]: Why `|| error`?

Comment (badrAZ) [reply src/index.js:7]: If the error.stack doesn't exist

File: src/index.js (hunk 3 above)
Comment (julien-f) [reply src/index.js]: `an URL` → `a URL`.

File: src/index.js (hunk 4 above)
Comment (julien-f) [reply src/index.js:7]: Because `error` can not have a `stack`.

File: README.md [hunk 5]
```diff
@@ -12,11 +12,64 @@ Installation of the [npm package](https://npmjs.org/package/xo-server-transport-
...
+
+### Slack
+
+#### Generate the Webhook
+
+1. Log in your Slack team account
+2. Click on the **Main menu** at the top and choose **Apps & Integrations**
+
+   ![Apps & Integrations](image/DocImg1.png)
+3. Search **Incoming WebHooks**
```
Comment (pdonias) [inline README.md:26]: Add an empty line between text and image.

File: README.md [hunk 6]
```diff
@@ -12,11 +12,64 @@ Installation of the [npm package](https://npmjs.org/package/xo-server-transport-
...
+   ![Incoming WebHooks](image/DocImg2.png)
+4. Click on **Add Configuration**
+
+   ![Add Configuration](image/DocImg3.png)
+5. Choose the default channel and click on **Add Incoming WebHooks integration**
```
Comment (pdonias) [inline README.md:32]: Idem.

File: README.md [hunk 7]
```diff
@@ -12,11 +12,64 @@ Installation of the [npm package](https://npmjs.org/package/xo-server-transport-
...
+   ![Add Incoming WebHooks integration](image/DocImg4.png)
+6. Modify the default settings and click on **Save Settings**
```
Comment (pdonias) [inline README.md:35]: Idem.

File: README.md (hunk 5 above)
Comment (julien-f) [reply README.md:26]: You mean between image and text?

Are we sure it won't break the list?

File: src/index.js [hunk 8]
```diff
@@ -1,27 +1,86 @@
...
-    throw new Error('not implemented')
+    return this._sendSlack({
+      text: `Hi there,
+
+The transport-slack plugin for Xen Orchestra server seems to be working fine, nicely done :)`
+    })
+  }
+
+  _sendSlack ({
+    text
```
Comment (julien-f) [inline src/index.js]: Use `message` instead of `text`.

File: src/index.js [hunk 9]
```diff
@@ -1,27 +1,86 @@
...
+  }) {
+    this._conf.text = text
```
Comment (julien-f) [inline src/index.js]: Nope, do not store the current message in the configuration.

File: src/index.js [hunk 10]
```diff
@@ -1,27 +1,86 @@
...
+    // TODO: handle errors
+    return this._send(this._conf).catch(logAndRethrow)
```
Comment (julien-f) [inline src/index.js]: `this._send({ ...this._conf, text: message })`

File: src/index.js [hunk 11]
```diff
@@ -1,27 +1,86 @@
...
+  additionalProperties: false,
+  required: ['webhookUri', 'channel']
+}
+
+// ===================================================================
+
+class XoServerTransportSlack {
   constructor ({ xo }) {
-    this._xo = xo
+    this._sendSlack = ::this._sendSlack
```
Comment (julien-f) [inline src/index.js:42]: Do you understand what this line does?

**Review by julien-f — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (pdonias) [reply README.md:26]: No, between text ("3. Search...") and image (DocImg2.png). Otherwise, the image is on the same line as the text.

Comment (julien-f) [reply README.md:26]: I see.

File: src/index.js (hunk 11 above)
Comment (badrAZ) [reply src/index.js:42]: When we call it, it will have as context this of the class.

File: README.md [hunk 12]
```diff
@@ -12,11 +12,67 @@ Installation of the [npm package](https://npmjs.org/package/xo-server-transport-
...
+
+This method is called to load the plugin.
+
+### `unload() `
+
+This method is called to unload the plugin.
+
+### `test() `
+
+This method is called to test the configuration plugin.
```
Comment (julien-f) [inline README.md]: Not necessary to document lifecycle methods, those are standard.

File: README.md [hunk 13]
```diff
@@ -12,11 +12,67 @@ Installation of the [npm package](https://npmjs.org/package/xo-server-transport-
...
+
+### `_sendSlack(message) `
```
Comment (julien-f) [inline README.md]: Document the exposed method instead: `Xo#sendSlackMessage({ message })`.

**Review by julien-f — CHANGES_REQUESTED** [review]

**Review by julien-f — APPROVED** [review]

**Pull request merged**